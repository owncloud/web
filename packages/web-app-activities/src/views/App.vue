<template>
  <h1 v-text="$gettext('Activities')" />
  <div class="oc-width-1-1 oc-mb-m">
    <item-filter
      ref="mediaTypeFilter"
      :allow-multiple="false"
      :filter-label="$gettext('Location')"
      :filterable-attributes="['name']"
      :option-filter-label="$gettext('Filter location')"
      :show-option-filter="true"
      :items="filterableSpaces"
      :close-on-click="true"
      class="files-search-filter-file-type oc-mr-s"
      display-name-attribute="name"
      filter-name="location"
    >
      <template #image="{ item }">
        <oc-icon :name="getLocationFilterIcon(item)" />
      </template>
      <template #item="{ item }">
        <div v-text="item.name" />
      </template>
    </item-filter>
  </div>
  <app-loading-spinner v-if="isLoading" />
  <template v-else>
    <no-content-message v-if="!activities.length" icon="pulse">
      <template #message>
        <span v-text="$gettext('No activities found')" />
      </template>
    </no-content-message>
    <ActivityList v-else :activities="activities" />
  </template>
</template>

<script lang="ts">
import { computed, defineComponent, onMounted, ref, unref, watch } from 'vue'
import {
  AppLoadingSpinner,
  ItemFilter,
  NoContentMessage,
  useClientService,
  useRouteQuery,
  useSpacesStore
} from '@ownclouders/web-pkg'
import { storeToRefs } from 'pinia'
import {
  call,
  isPersonalSpaceResource,
  isProjectSpaceResource,
  SpaceResource
} from '@ownclouders/web-client'
import { useTask } from 'vue-concurrency'
import { Activity } from '@ownclouders/web-client/graph/generated'
import ActivityList from './ActivityList.vue'

export default defineComponent({
  name: 'App',
  components: { ActivityList, NoContentMessage, ItemFilter, AppLoadingSpinner },
  setup() {
    const spacesStore = useSpacesStore()
    const { spaces } = storeToRefs(spacesStore)
    const clientService = useClientService()
    const activities = ref<Activity[]>([])

    const locationQuery = useRouteQuery('q_location')

    const filterableSpaces = computed(() => {
      return [...unref(spaces)]
        .filter(
          (space) =>
            !space.disabled && (isProjectSpaceResource(space) || isPersonalSpaceResource(space))
        )
        .sort((a, b) => {
          if (isPersonalSpaceResource(a) === isPersonalSpaceResource(b)) {
            return a.name.localeCompare(b.name)
          }
          return isPersonalSpaceResource(a) ? -1 : 1
        })
    })

    const loadActivitiesTask = useTask(function* (signal) {
      const filters = ['sort:desc', 'limit:100']

      if (unref(locationQuery)) {
        filters.push(`itemid:${unref(locationQuery)}`)
      }

      activities.value = yield* call(
        clientService.graphAuthenticated.activities.listActivities(filters.join(' AND '), {
          signal
        })
      )
    })

    const isLoading = computed(() => loadActivitiesTask.isRunning || !loadActivitiesTask.last)

    const getLocationFilterIcon = (space: SpaceResource) => {
      if (isPersonalSpaceResource(space)) {
        return 'folder'
      }

      return 'layout-grid'
    }

    onMounted(() => {
      loadActivitiesTask.perform()
    })

    watch(locationQuery, () => {
      loadActivitiesTask.perform()
    })

    return { activities, filterableSpaces, getLocationFilterIcon, isLoading }
  }
})
</script>
