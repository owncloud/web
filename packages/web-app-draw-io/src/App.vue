<template>
  <iframe
    id="drawio-editor"
    ref="drawIoEditor"
    :src="iframeSource"
    :title="$gettext('Draw.io editor')"
  />
</template>

<script lang="ts">
import qs from 'qs'
import {
  defineComponent,
  unref,
  PropType,
  computed,
  onMounted,
  onBeforeUnmount,
  VNodeRef,
  ref,
  watch,
  nextTick
} from 'vue'
import { Resource } from '@ownclouders/web-client'
import { AppConfigObject } from '@ownclouders/web-pkg'

export default defineComponent({
  name: 'DrawIoEditor',
  props: {
    resource: {
      type: Object as PropType<Resource>,
      required: true
    },
    applicationConfig: {
      type: Object as PropType<AppConfigObject>,
      required: true
    },
    currentContent: {
      type: String,
      required: true
    },
    isReadOnly: {
      type: Boolean,
      required: true
    },
    isDirty: {
      type: Boolean,
      required: true
    }
  },
  emits: ['update:currentContent', 'save', 'close'],
  setup(props, { emit }) {
    const drawIoEditor: VNodeRef = ref()

    const config = computed(() => {
      const { url = 'https://embed.diagrams.net', theme = 'minimal' } =
        props.applicationConfig as any
      return { url, theme }
    })

    const urlHost = computed(() => {
      const url = new URL(unref(config).url)
      const urlHost = `${url.protocol}//${url.hostname}`
      return url.port ? `${urlHost}:${url.port}` : urlHost
    })

    const iframeSource = computed(() => {
      const query = qs.stringify({
        embed: 1,
        chrome: props.isReadOnly ? 0 : 1,
        picker: 0,
        stealth: 1,
        spin: 1,
        proto: 'json',
        ui: unref(config).theme
      })

      return `${unref(config).url}?${query}`
    })

    const loadCurrentContent = () => {
      postMessage({
        action: 'load',
        xml: props.currentContent,
        autosave: true
      })
    }

    watch(
      () => props.isDirty,
      () => {
        postMessage({
          action: 'status',
          modified: props.isDirty
        })
      }
    )

    watch(
      () => props.resource,
      () => {
        loadCurrentContent()
      }
    )

    const postMessage = (
      payload:
        | { action: string; xml: string; autosave?: boolean }
        | { action: 'status'; modified: boolean }
    ) => {
      try {
        if (!unref(drawIoEditor)) {
          return
        }
        return (unref(drawIoEditor) as HTMLIFrameElement).contentWindow.postMessage(
          JSON.stringify(payload),
          unref(urlHost)
        )
      } catch (e) {
        console.error(e)
      }
    }

    const handleMessage = async (event: MessageEvent) => {
      if (event.data.length > 0) {
        if (event.origin !== unref(config).url) {
          return
        }
        await nextTick()
        const payload = JSON.parse(event.data)
        switch (payload?.event) {
          case 'init':
            loadCurrentContent()
            break
          case 'autosave':
            emit('update:currentContent', payload.xml)
            break
          case 'save':
            emit('save')
            break
          case 'exit':
            emit('close')
            break
        }
      }
    }

    onMounted(() => {
      window.addEventListener('message', handleMessage)
    })

    onBeforeUnmount(() => {
      window.removeEventListener('message', handleMessage)
    })

    return {
      config,
      drawIoEditor,
      iframeSource
    }
  }
})
</script>
<style scoped>
#drawio-editor {
  width: 100%;
  height: 100%;
  border: none;
  margin: 0;
  padding: 0;
  overflow: hidden;
}
</style>
