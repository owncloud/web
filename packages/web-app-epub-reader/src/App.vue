<template>
  <div class="epub-reader oc-flex">
    <oc-list class="epub-reader-chapters-list oc-pl-s oc-visible@l">
      <li
        v-for="chapter in chapters"
        :key="chapter.id"
        class="epub-reader-chapters-list-item oc-py-s"
        :class="{ active: currentChapter.id === chapter.id }"
      >
        <oc-button class="oc-text-truncate" appearance="raw" @click="showChapter(chapter)">
          <span
            v-oc-tooltip="chapter.label"
            class="oc-text-truncate oc-mr-s"
            v-text="chapter.label"
          />
        </oc-button>
      </li>
    </oc-list>
    <div class="oc-width-1-1 oc-height-1-1">
      <div class="epub-reader-controls oc-flex oc-flex-middle oc-m-s">
        <div class="epub-reader-controls-font-size oc-flex oc-button-group">
          <oc-button
            v-oc-tooltip="`${currentFontSizePercentage - FONT_SIZE_PERCENTAGE_STEP}%`"
            :aria-label="$gettext('Decrease font size')"
            class="epub-reader-controls-font-size-decrease"
            :disabled="decreaseFontSizeDisabled"
            gap-size="none"
            @click="decreaseFontSize"
          >
            <oc-icon name="font-family" fill-type="none" size="small" />
            <oc-icon name="subtract" size="xsmall" />
          </oc-button>
          <oc-button
            v-oc-tooltip="$gettext('Reset font size')"
            class="epub-reader-controls-font-size-reset"
            gap-size="none"
            @click="resetFontSize"
            v-text="`${currentFontSizePercentage}%`"
          />
          <oc-button
            v-oc-tooltip="`${currentFontSizePercentage + FONT_SIZE_PERCENTAGE_STEP}%`"
            :aria-label="$gettext('Increase font size')"
            class="epub-reader-controls-font-size-increase"
            :disabled="increaseFontSizeDisabled"
            gap-size="none"
            @click="increaseFontSize"
          >
            <oc-icon name="font-family" fill-type="none" size="small" />
            <oc-icon name="add" size="xsmall" />
          </oc-button>
        </div>
        <oc-select
          v-model="currentChapter"
          class="epub-reader-controls-chapters-select oc-width-1-1 oc-px-s oc-hidden@l"
          :label="$gettext('Chapter')"
          :label-hidden="true"
          :options="chapters"
          :searchable="false"
          @update:model-value="showChapter"
        />
      </div>
      <div class="oc-flex oc-flex-center oc-width-1-1 oc-height-1-1">
        <div class="oc-flex oc-flex-middle oc-mx-l">
          <oc-button
            class="epub-reader-navigate-left"
            :aria-label="$gettext('Navigate to previous page')"
            :disabled="navigateLeftDisabled"
            appearance="raw"
            @click="navigateLeft"
          >
            <oc-icon name="arrow-left-s" fill-type="line" size="xlarge" />
          </oc-button>
        </div>
        <div id="reader" ref="bookContainer" class="oc-flex oc-flex-center" />

        <div class="oc-flex oc-flex-middle oc-mx-l">
          <oc-button
            class="epub-reader-navigate-right"
            :aria-label="$gettext('Navigate to next page')"
            :disabled="navigateRightDisabled"
            appearance="raw"
            @click="navigateRight"
          >
            <oc-icon name="arrow-right-s" fill-type="line" size="xlarge" />
          </oc-button>
        </div>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { computed, defineComponent, nextTick, PropType, ref, unref, watch } from 'vue'
import { Resource } from '@ownclouders/web-client'
import {
  AppConfigObject,
  Key,
  useKeyboardActions,
  useLocalStorage,
  useThemeStore
} from '@ownclouders/web-pkg'
import ePub, { Book, NavItem, Rendition, Location } from 'epubjs'

const DARK_THEME_CONFIG = {
  html: {
    '-webkit-filter': 'invert(1) hue-rotate(180deg)',
    filter: 'invert(1) hue-rotate(180deg)'
  },
  img: {
    '-webkit-filter': 'invert(1) hue-rotate(180deg)',
    filter: 'invert(1) hue-rotate(180deg)'
  }
}
const LIGHT_THEME_CONFIG = {
  html: { background: 'white' }
}
const MAX_FONT_SIZE_PERCENTAGE = 150
const MIN_FONT_SIZE_PERCENTAGE = 50
const FONT_SIZE_PERCENTAGE_STEP = 10

export default defineComponent({
  name: 'EpubReader',
  props: {
    applicationConfig: { type: Object as PropType<AppConfigObject>, required: true },
    currentContent: {
      type: String,
      required: true
    },
    isReadOnly: { type: Boolean, required: false },
    resource: { type: Object as PropType<Resource>, required: true }
  },
  emits: ['close'],
  setup(props) {
    const keyboardActions = useKeyboardActions()
    const bookContainer = ref<Element>()
    const chapters = ref<NavItem[]>([])
    const currentChapter = ref<NavItem>()
    const navigateLeftDisabled = ref(false)
    const navigateRightDisabled = ref(false)
    const localStorageData = useLocalStorage<{ fontSizePercentage?: number }>(`oc_epubReader`, {})
    const currentFontSizePercentage = ref(unref(localStorageData).fontSizePercentage || 100)
    const themeStore = useThemeStore()
    let book = ref<Book>()
    let rendition = ref<Rendition>()

    const navigateLeft = () => {
      unref(rendition).prev()
    }

    const navigateRight = () => {
      unref(rendition).next()
    }

    const showChapter = (chapter: NavItem) => {
      unref(rendition).display(chapter.href)
    }

    const increaseFontSize = () => {
      currentFontSizePercentage.value = Math.min(
        unref(currentFontSizePercentage) + FONT_SIZE_PERCENTAGE_STEP,
        MAX_FONT_SIZE_PERCENTAGE
      )
    }

    const resetFontSize = () => {
      currentFontSizePercentage.value = 100
    }

    const decreaseFontSize = () => {
      currentFontSizePercentage.value = Math.max(
        unref(currentFontSizePercentage) - FONT_SIZE_PERCENTAGE_STEP,
        MIN_FONT_SIZE_PERCENTAGE
      )
    }

    const increaseFontSizeDisabled = computed(() => {
      return unref(currentFontSizePercentage) >= MAX_FONT_SIZE_PERCENTAGE
    })

    const decreaseFontSizeDisabled = computed(() => {
      return unref(currentFontSizePercentage) <= MIN_FONT_SIZE_PERCENTAGE
    })

    keyboardActions.bindKeyAction({ primary: Key.ArrowLeft }, () => navigateLeft())
    keyboardActions.bindKeyAction({ primary: Key.ArrowRight }, () => navigateRight())

    watch(
      () => props.currentContent,
      async () => {
        await nextTick()

        if (unref(book)) {
          unref(book).destroy()
        }

        const localStorageResourceData = useLocalStorage<{ currentLocation?: Location }>(
          `oc_epubReader_resource_${props.resource.id}`,
          {}
        )

        book.value = ePub(props.currentContent)

        unref(book).loaded.navigation.then(({ toc }) => {
          chapters.value = toc
          currentChapter.value = toc?.[0]
        })

        rendition.value = unref(book).renderTo(unref(bookContainer), {
          flow: 'paginated',
          width: 650,
          height: '90%' // Don't use full height to avoid cut-off text
        })

        unref(rendition).themes.register('dark', DARK_THEME_CONFIG)
        unref(rendition).themes.register('light', LIGHT_THEME_CONFIG)
        unref(rendition).themes.select(themeStore.currentTheme.isDark ? 'dark' : 'light')
        unref(rendition).themes.fontSize(`${unref(currentFontSizePercentage)}%`)
        unref(rendition).display(unref(localStorageResourceData)?.currentLocation?.start?.cfi)

        unref(rendition).on('keydown', (event: KeyboardEvent) => {
          if (event.key === Key.ArrowLeft) {
            navigateLeft()
          }
          if (event.key === Key.ArrowRight) {
            navigateRight()
          }
        })

        unref(rendition).on('relocated', () => {
          const currentLocation = unref(rendition).currentLocation() as any & Location
          localStorageResourceData.value = { currentLocation }
          navigateLeftDisabled.value = currentLocation.atStart === true
          navigateRightDisabled.value = currentLocation.atEnd === true

          const locationCfi = currentLocation.start.cfi
          const spineItem = unref(book).spine.get(locationCfi)
          const navItem = unref(book).navigation.get(spineItem.href)
          // Might be sub nav item and therefore undefined
          if (navItem) {
            currentChapter.value = navItem
          }
        })
      },
      {
        immediate: true
      }
    )

    watch(currentFontSizePercentage, () => {
      unref(rendition).themes.fontSize(`${unref(currentFontSizePercentage)}%`)
      localStorageData.value = {
        ...unref(localStorageData),
        fontSizePercentage: unref(currentFontSizePercentage)
      }
    })

    return {
      bookContainer,
      navigateLeft,
      navigateLeftDisabled,
      navigateRight,
      navigateRightDisabled,
      currentChapter,
      chapters,
      showChapter,
      resetFontSize,
      increaseFontSize,
      decreaseFontSize,
      increaseFontSizeDisabled,
      decreaseFontSizeDisabled,
      currentFontSizePercentage,
      FONT_SIZE_PERCENTAGE_STEP,
      rendition,
      book
    }
  }
})
</script>
<style lang="scss">
.epub-reader {
  &-chapters-list {
    background: var(--oc-color-background-muted);
    border-right: 1px solid var(--oc-color-border);
    width: 240px;
    overflow-y: auto;

    &-item:not(:last-child) {
      border-bottom: 1px solid var(--oc-color-border);
    }

    &-item.active {
      .oc-button {
        color: var(--oc-color-swatch-primary-default);
      }
    }
  }

  &-controls-font-size {
    flex-wrap: nowrap;

    &-reset {
      width: 58px; //prevent jumpy behaviour
    }
  }
}
</style>
