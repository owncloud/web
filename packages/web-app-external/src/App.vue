<template>
  <iframe
    v-if="appUrl && method === 'GET'"
    :src="appUrl"
    class="oc-width-1-1 oc-height-1-1"
    :title="iFrameTitle"
    allowfullscreen
  />
  <div v-if="appUrl && method === 'POST' && formParameters" class="oc-height-1-1 oc-width-1-1">
    <form :action="appUrl" target="app-iframe" method="post">
      <input ref="subm" type="submit" :value="formParameters" class="oc-hidden" />
      <div v-for="(item, key, index) in formParameters" :key="index">
        <input :name="key" :value="item" type="hidden" />
      </div>
    </form>
    <iframe
      name="app-iframe"
      class="oc-width-1-1 oc-height-1-1"
      :title="iFrameTitle"
      allowfullscreen
    />
  </div>
</template>

<script lang="ts">
import { stringify } from 'qs'
import {
  PropType,
  computed,
  defineComponent,
  unref,
  nextTick,
  ref,
  watch,
  VNodeRef,
  onMounted
} from 'vue'
import { useTask } from 'vue-concurrency'
import { useGettext } from 'vue3-gettext'

import { Resource, SpaceResource } from '@ownclouders/web-client'
import { urlJoin } from '@ownclouders/web-client'
import {
  isSameResource,
  useCapabilityStore,
  useConfigStore,
  useMessages,
  useRequest,
  useAppProviderService,
  useRoute
} from '@ownclouders/web-pkg'
import {
  isProjectSpaceResource,
  isPublicSpaceResource,
  isShareSpaceResource
} from '@ownclouders/web-client'

export default defineComponent({
  name: 'ExternalApp',
  props: {
    space: { type: Object as PropType<SpaceResource>, required: true },
    resource: { type: Object as PropType<Resource>, required: true },
    isReadOnly: { type: Boolean, required: true }
  },
  setup(props) {
    const language = useGettext()
    const { $gettext } = language
    const { showErrorMessage } = useMessages()
    const capabilityStore = useCapabilityStore()
    const configStore = useConfigStore()
    const route = useRoute()
    const appProviderService = useAppProviderService()
    const { makeRequest } = useRequest()

    const appName = computed(() => {
      const lowerCaseAppName = unref(route)
        .name.toString()
        .replace('external-', '')
        .replace('-apps', '')
      return appProviderService.appNames.find(
        (appName) => appName.toLowerCase() === lowerCaseAppName
      )
    })

    const appUrl = ref()
    const formParameters = ref({})
    const method = ref()
    const subm: VNodeRef = ref()

    const iFrameTitle = computed(() => {
      return $gettext('"%{appName}" app content area', {
        appName: unref(appName)
      })
    })

    const errorPopup = (error: string) => {
      showErrorMessage({
        title: $gettext('An error occurred'),
        desc: error,
        errors: [new Error(error)]
      })
    }

    const loadAppUrl = useTask(function* (signal, viewMode: string) {
      try {
        if (props.isReadOnly && viewMode === 'write') {
          showErrorMessage({ title: $gettext('Cannot open file in edit mode as it is read-only') })
          return
        }

        const fileId = props.resource.fileId
        const baseUrl = urlJoin(
          configStore.serverUrl,
          capabilityStore.filesAppProviders[0].open_url
        )

        const query = stringify({
          file_id: fileId,
          lang: language.current,
          ...(unref(appName) && { app_name: encodeURIComponent(unref(appName)) }),
          ...(viewMode && { view_mode: viewMode })
        })

        const url = `${baseUrl}?${query}`
        const response = yield makeRequest('POST', url, {
          validateStatus: () => true
        })

        if (response.status !== 200) {
          switch (response.status) {
            case 425:
              errorPopup(
                $gettext(
                  'This file is currently being processed and is not yet available for use. Please try again shortly.'
                )
              )
              break
            default:
              errorPopup(response.data?.message)
          }

          throw new Error('Error fetching app information')
        }

        if (!response.data.app_url || !response.data.method) {
          throw new Error('Error in app server response')
        }

        appUrl.value = response.data.app_url
        method.value = response.data.method

        if (response.data.form_parameters) {
          formParameters.value = response.data.form_parameters
        }

        if (method.value === 'POST' && formParameters.value) {
          // eslint-disable-next-line vue/valid-next-tick
          yield nextTick()
          unref(subm).click()
        }
      } catch (e) {
        console.error('web-app-external error', e)
        throw e
      }
    }).restartable()

    const determineOpenAsPreview = (appName: string) => {
      const openAsPreview = configStore.options.editor.openAsPreview
      return (
        openAsPreview === true || (Array.isArray(openAsPreview) && openAsPreview.includes(appName))
      )
    }

    // switch to write mode when edit is clicked
    const catchClickMicrosoftEdit = (event: MessageEvent) => {
      try {
        if (JSON.parse(event.data)?.MessageId === 'UI_Edit') {
          loadAppUrl.perform('write')
        }
      } catch (e) {}
    }
    onMounted(() => {
      if (determineOpenAsPreview(unref(appName))) {
        window.addEventListener('message', catchClickMicrosoftEdit)
      } else {
        window.removeEventListener('message', catchClickMicrosoftEdit)
      }
    })

    watch(
      [props.resource],
      ([newResource], [oldResource]) => {
        if (isSameResource(newResource, oldResource)) {
          return
        }

        let viewMode = props.isReadOnly ? 'view' : 'write'
        if (
          determineOpenAsPreview(unref(appName)) &&
          (isShareSpaceResource(props.space) ||
            isPublicSpaceResource(props.space) ||
            isProjectSpaceResource(props.space))
        ) {
          viewMode = 'view'
        }
        loadAppUrl.perform(viewMode)
      },
      { immediate: true, deep: true }
    )

    return {
      appUrl,
      formParameters,
      iFrameTitle,
      method,
      subm
    }
  }
})
</script>
