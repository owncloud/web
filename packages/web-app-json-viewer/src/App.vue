<template>
  <div class="oc-json-viewer oc-width-1-1 oc-height-1-1">
    <div
      ref="editorRef"
      class="oc-width-1-1 oc-height-1-1"
      :class="{ 'jse-theme-dark': darkTheme }"
    ></div>
  </div>
</template>

<script lang="ts">
import { computed, defineComponent, onMounted, PropType, ref, unref } from 'vue'
import { Resource } from '@ownclouders/web-client'
import { AppConfigObject, useMessages, useThemeStore } from '@ownclouders/web-pkg'
import { Content, JSONEditor, Mode } from 'vanilla-jsoneditor'
import { useGettext } from 'vue3-gettext'
import 'vanilla-jsoneditor/themes/jse-theme-dark.css'

export default defineComponent({
  props: {
    applicationConfig: { type: Object as PropType<AppConfigObject>, required: true },
    currentContent: {
      type: String,
      required: true
    },
    isReadOnly: { type: Boolean, required: false },
    resource: { type: Object as PropType<Resource>, required: true }
  },
  setup: (props) => {
    const editorRef = ref<HTMLElement>()
    const { $gettext } = useGettext()
    const { showErrorMessage } = useMessages()
    const themeStore = useThemeStore()
    const darkTheme = computed(() => {
      return themeStore.currentTheme.isDark
    })

    let content: Content = null
    try {
      content = {
        json: JSON.parse(props.currentContent),
        text: props.currentContent
      }
    } catch (e) {
      content = {
        json: undefined,
        text: props.currentContent
      }
      showErrorMessage({ title: $gettext('JSON not properly formatted') })
    }

    onMounted(async () => {
      const editor = new JSONEditor({
        target: unref(editorRef),
        props: {
          content,
          mode: Mode.tree,
          askToFormat: false,
          readOnly: true,
          navigationBar: false
        }
      })

      // expand all paths up to 2 levels
      await editor.expand((path) => path.length < 2)
    })

    return {
      editorRef,
      darkTheme
    }
  }
})
</script>
<style lang="scss">
.jse {
  &-menu {
    background-color: var(--oc-color-swatch-primary-gradient) !important;
  }

  &-error {
    display: none !important; // Hide, as we show our own error via showErrorMessage to keep it simple
  }

  &-menu {
    button:nth-of-type(1) {
      border-bottom-left-radius: 8px !important;
      border-top-left-radius: 8px !important;
    }
    button:nth-of-type(3) {
      border-bottom-right-radius: 8px !important;
      border-top-right-radius: 8px !important;
    }
  }
}
</style>
