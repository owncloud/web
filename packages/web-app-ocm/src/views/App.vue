<template>
  <div class="sciencemesh">
    <div class="oc-flex oc-flex-column sciencemesh-wrapper">
      <div class="oc-flex sciencemesh-top">
        <div id="sciencemesh-invite" class="oc-width-1-2">
          <outgoing-invitations />
        </div>
        <div id="sciencemesh-accept-invites" class="oc-width-1-2">
          <incoming-invitations @highlight-new-connections="highlightNewConnections" />
        </div>
      </div>
      <div id="sciencemesh-connections">
        <connections-panel
          v-model:connections="connections"
          :highlighted-connections="highlightedConnections.map((c) => c.id)"
          :loading="loadingConnections"
        />
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent, onMounted, onUnmounted, ref, unref, Ref } from 'vue'
import ConnectionsPanel from './ConnectionsPanel.vue'
import IncomingInvitations from './IncomingInvitations.vue'
import OutgoingInvitations from './OutgoingInvitations.vue'
import {
  useClientService,
  useScrollTo,
  FederatedConnection,
  useMessages
} from '@ownclouders/web-pkg'
import { useGettext } from 'vue3-gettext'
import { buildConnection } from '../functions'

export default defineComponent({
  components: {
    IncomingInvitations,
    OutgoingInvitations,
    ConnectionsPanel
  },
  setup() {
    const { showMessage } = useMessages()
    const { scrollToResource } = useScrollTo()
    const clientSerivce = useClientService()
    const { $gettext } = useGettext()

    const connections: Ref<FederatedConnection[]> = ref([])
    const highlightedConnections: Ref<FederatedConnection[]> = ref([])
    const highlightNewConnectionsInterval = ref(null)
    const loadingConnections = ref(true)

    const findAcceptedUsers = async () => {
      try {
        const { data: acceptedUsers } = await clientSerivce.httpAuthenticated.get<
          FederatedConnection[]
        >('/sciencemesh/find-accepted-users')
        loadingConnections.value = false
        connections.value = acceptedUsers.map(buildConnection)
      } catch {
        connections.value = []
        loadingConnections.value = false
      }
    }

    const highlightNewConnections = async () => {
      const oldConnections = [...unref(connections)]
      await findAcceptedUsers()
      if (oldConnections.length < unref(connections).length) {
        highlightedConnections.value = unref(connections).filter(
          (connection) => !oldConnections.map((c) => c.id).includes(connection.id)
        )
        if (unref(highlightedConnections).length === 1) {
          scrollToResource(unref(highlightedConnections)[0].id)
          showMessage({
            title: $gettext('New federated connection'),
            status: 'success',
            desc: $gettext('You can share with and recieve shares from %{user} now', {
              user: unref(highlightedConnections)[0].display_name
            })
          })
        } else if (unref(highlightedConnections).length > 1) {
          const newConnections = unref(highlightedConnections)
            .map((c) => c.display_name)
            .join(', ')

          showMessage({
            title: $gettext('New federated connections'),
            status: 'success',
            desc: $gettext('You can share with and receive shares from %{ connections } now', {
              connections: newConnections
            })
          })
        }
      }
    }

    onMounted(async () => {
      await findAcceptedUsers()
      loadingConnections.value = false
      highlightNewConnectionsInterval.value = setInterval(() => {
        highlightNewConnections()
      }, 10 * 1000)
    })

    onUnmounted(() => {
      clearInterval(unref(highlightNewConnectionsInterval))
    })

    return {
      highlightNewConnections,
      connections,
      highlightedConnections,
      loadingConnections
    }
  }
})
</script>

<style lang="scss" scoped>
.sciencemesh {
  background-color: var(--oc-color-background-hover);
  overflow: auto;
}
.sciencemesh-wrapper {
  height: 100%;
}
.sciencemesh-top {
  max-height: 360px;
  @media (max-width: $oc-breakpoint-large-default) {
    flex-direction: column;
    height: auto;
    max-height: unset;
  }
}
#sciencemesh-invite,
#sciencemesh-accept-invites {
  margin: var(--oc-space-small);
  overflow: auto;
}
#sciencemesh-invite,
#sciencemesh-accept-invites,
#sciencemesh-connections {
  margin: var(--oc-space-small);
  background-color: var(--oc-color-background-default);
  border-radius: 15px;
  padding: var(--oc-space-small);
  @media (max-width: $oc-breakpoint-large-default) {
    width: auto;
  }
}
#sciencemesh-connections {
  flex: 1;
}
#sciencemesh-invite {
  overflow: auto;
  @media (max-width: $oc-breakpoint-large-default) {
    margin-bottom: var(--oc-space-small);
  }
}
</style>
