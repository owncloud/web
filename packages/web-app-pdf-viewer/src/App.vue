<template>
  <object
    class="pdf-viewer oc-width-1-1 oc-height-1-1"
    :data="url"
    :type="objectType"
    :aria-label="
      $pgettext('Accessible label for the object holding the PDF file content', 'PDF document')
    "
  />
</template>

<script lang="ts" setup>
interface Props {
  url: string
}
const { url } = defineProps<Props>()
const isSafari = navigator.userAgent?.includes('Safari') && !navigator.userAgent?.includes('Chrome')
const objectType = isSafari ? undefined : 'application/pdf'
</script>

<style scoped>
.pdf-viewer {
  border: none;
  margin: 0;
  padding: 0;
  overflow: hidden;
}
</style>
