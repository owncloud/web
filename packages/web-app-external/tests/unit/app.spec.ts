import { mock } from 'vitest-mock-extended'
import { defaultComponentMocks, defaultPlugins, shallowMount } from 'web-test-helpers'
import { AppProviderService, useRequest, useRoute } from '@ownclouders/web-pkg'
import { ref } from 'vue'

import { Resource } from '@ownclouders/web-client'
import App from '../../src/App.vue'
import { RouteLocation } from 'vue-router'

vi.mock('@ownclouders/web-pkg', async (importOriginal) => ({
  ...(await importOriginal<any>()),
  useRequest: vi.fn(),
  useRoute: vi.fn()
}))

const appUrl = 'https://example.test/d12ab86/loe009157-MzBw'

const providerSuccessResponsePost = {
  app_url: appUrl,
  method: 'POST',
  form_parameters: {
    access_token: 'asdfsadfsadf',
    access_token_ttl: '123456'
  }
}

const providerSuccessResponseGet = {
  app_url: appUrl,
  method: 'GET'
}

describe('The app provider extension', () => {
  beforeEach(() => {
    vi.spyOn(console, 'error').mockImplementation(() => undefined)
  })

  it('should fail for unauthenticated users', async () => {
    const makeRequest = vi.fn().mockResolvedValue({
      ok: true,
      status: 401,
      message: 'Login Required'
    })
    const { wrapper } = createShallowMountWrapper(makeRequest)
    await wrapper.vm.$nextTick()
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('should be able to load an iFrame via get', async () => {
    const makeRequest = vi.fn().mockResolvedValue({
      ok: true,
      status: 200,
      data: providerSuccessResponseGet
    })

    const { wrapper } = createShallowMountWrapper(makeRequest)
    await wrapper.vm.$nextTick()
    await wrapper.vm.$nextTick()
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('should be able to load an iFrame via post', async () => {
    const makeRequest = vi.fn().mockResolvedValue({
      ok: true,
      status: 200,
      data: providerSuccessResponsePost
    })
    const { wrapper } = createShallowMountWrapper(makeRequest)
    await wrapper.vm.$nextTick()
    await wrapper.vm.$nextTick()
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })
})

function createShallowMountWrapper(makeRequest = vi.fn().mockResolvedValue({ status: 200 })) {
  vi.mocked(useRequest).mockImplementation(() => ({
    makeRequest
  }))
  vi.mocked(useRoute).mockImplementation(() =>
    ref(mock<RouteLocation>({ name: 'external-example-app-apps' }))
  )
  const mocks = {
    ...defaultComponentMocks(),
    $appProviderService: mock<AppProviderService>({ appNames: ['example-app'] })
  }

  const capabilities = {
    files: {
      app_providers: [{ apps_url: '/app/list', enabled: true, open_url: '/app/open' }]
    }
  }

  return {
    wrapper: shallowMount(App, {
      props: {
        space: null,
        resource: mock<Resource>(),
        isReadOnly: false
      },
      global: {
        plugins: [
          ...defaultPlugins({
            piniaOptions: {
              capabilityState: { capabilities },
              configState: { options: { editor: { openAsPreview: true } } }
            }
          })
        ],
        provide: mocks,
        mocks
      }
    })
  }
}
