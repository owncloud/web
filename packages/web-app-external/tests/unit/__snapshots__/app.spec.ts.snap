// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`The app provider extension > should be able to load an iFrame via get 1`] = `
"<iframe src="https://example.test/d12ab86/loe009157-MzBw" class="oc-width-1-1 oc-height-1-1" title="" example-app" app content area" allowfullscreen=""></iframe>

<!--v-if-->"
`;

exports[`The app provider extension > should be able to load an iFrame via post 1`] = `
"<!--v-if-->

<div class="oc-height-1-1 oc-width-1-1">
  <form action="https://example.test/d12ab86/loe009157-MzBw" target="app-iframe" method="post"><input type="submit" class="oc-hidden" value="[object Object]">
    <div><input name="access_token" type="hidden" value="asdfsadfsadf"></div>
    <div><input name="access_token_ttl" type="hidden" value="123456"></div>
  </form> <iframe name="app-iframe" class="oc-width-1-1 oc-height-1-1" title="" example-app" app content area" allowfullscreen=""></iframe>
</div>"
`;

exports[`The app provider extension > should fail for unauthenticated users 1`] = `
"<!--v-if-->

<!--v-if-->"
`;
